salt/utils/aws.py

```diff
@@ -71,6 +90,44 @@ def sleep_exponential_backoff(attempts):
     time.sleep(random.uniform(1, 2**attempts))
 
 
+def get_metadata(path, refresh_token_if_needed=True):
+    """
+    Get the instance metadata at the provided path
+    The path argument will be prepended by http://169.254.169.254/latest/
+    If using IMDSv2 with tokens required, the token will be fetched and used for subsequent requests
+    (unless refresh_token_if_needed is False, in which case this will fail if tokens are required
+    and no token was already cached)
+    """
+    global __IMDS_Token__
+
+    headers = {}
+    if __IMDS_Token__ is not None:
+        headers["X-aws-ec2-metadata-token"] = __IMDS_Token__
+
+    # Connections to instance meta-data must fail fast and never be proxied
+    result = requests.get(
+        "http://169.254.169.254/latest/{}".format(path),
+        proxies={"http": ""},
+        headers=headers,
+        timeout=AWS_METADATA_TIMEOUT,
+    )
+
+    if result.status_code == 401 and refresh_token_if_needed:
```

Comment:
Is there a way to determine someone is using IMDSv2 without just expecting a failure?

Reply:
Potentially, but it would still require an HTTP request, and having to run that check before each request would be slower than just handling the failure when it happens. Current method also handles token expiration gracefully without any extra logic needed.